File: frameworks/ml/nn/runtime/test/fuzzing/operation_signatures/Resize.cpp
/*
 * Copyright (C) 2019 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "fuzzing/operation_signatures/OperationSignatureUtils.h"

namespace android {
namespace nn {
namespace fuzzing_test {

static void resizeOpConstructor(Type, uint32_t rank, RandomOperation* op) {
    NN_FUZZER_CHECK(rank == 4);
    setFreeDimensions(op->inputs[0], rank);

    bool useNchw = op->inputs.size() > 3 ? static_cast<bool>(op->inputs[3]->value<bool8>()) : false;
    int heightIndex = useNchw ? 2 : 1;
    int widthIndex = useNchw ? 3 : 2;

    // Compute output height and width.
    RandomVariable outHeight, outWidth;
    switch (op->inputs[1]->dataType) {
        // Resize by shape.
        case Type::INT32:
            outWidth = op->inputs[1]->value<RandomVariable>();
            outHeight = op->inputs[2]->value<RandomVariable>();
            break;
        // Resize by scale.
        case Type::FLOAT32:
            outWidth = op->inputs[0]->dimensions[widthIndex] * op->inputs[1]->value<float>();
            outHeight = op->inputs[0]->dimensions[heightIndex] * op->inputs[2]->value<float>();
            break;
        case Type::FLOAT16:
            outWidth = op->inputs[0]->dimensions[widthIndex] *
                       static_cast<float>(op->inputs[1]->value<_Float16>());
            outHeight = op->inputs[0]->dimensions[heightIndex] *
                        static_cast<float>(op->inputs[2]->value<_Float16>());
            break;
        default:
            NN_FUZZER_CHECK(false) << "Unsupported data type.";
    }

    if (useNchw) {
        op->outputs[0]->dimensions = {op->inputs[0]->dimensions[0], op->inputs[0]->dimensions[1],
                                      outHeight, outWidth};
    } else {
        op->outputs[0]->dimensions = {op->inputs[0]->dimensions[0], outHeight, outWidth,
                                      op->inputs[0]->dimensions[3]};
    }
    setSameQuantization(op->outputs[0], op->inputs[0]);
}

#define DEFINE_RESIZE_WITHOUT_LAYOUT_SIGNATURE(op, ver, ...)             \
    DEFINE_OPERATION_SIGNATURE(op##_no_layout_##ver){                    \
            .opType = ANEURALNETWORKS_##op,                              \
            .supportedDataTypes = {__VA_ARGS__},                         \
            .supportedRanks = {4},                                       \
            .version = HalVersion::ver,                                  \
            .inputs = {INPUT_DEFAULT, RANDOM_INT_FREE, RANDOM_INT_FREE}, \
            .outputs = {OUTPUT_DEFAULT},                                 \
            .constructor = resizeOpConstructor};

DEFINE_RESIZE_WITHOUT_LAYOUT_SIGNATURE(RESIZE_BILINEAR, V1_0, Type::TENSOR_FLOAT32);
DEFINE_RESIZE_WITHOUT_LAYOUT_SIGNATURE(RESIZE_BILINEAR, V1_2, Type::TENSOR_QUANT8_ASYMM,
                                       Type::TENSOR_FLOAT16);

#define DEFINE_RESIZE_OP_SIGNATURE(op, ver, ...)                        \
    DEFINE_OPERATION_SIGNATURE(op##_shape_##ver){                       \
            .opType = ANEURALNETWORKS_##op,                             \
            .supportedDataTypes = {__VA_ARGS__},                        \
            .supportedRanks = {4},                                      \
            .version = HalVersion::ver,                                 \
            .inputs = {INPUT_DEFAULT, RANDOM_INT_FREE, RANDOM_INT_FREE, \
                       PARAMETER_CHOICE(Type::BOOL, false, true)},      \
            .outputs = {OUTPUT_DEFAULT},                                \
            .constructor = resizeOpConstructor};                        \
    DEFINE_OPERATION_SIGNATURE(op##_scale_##ver){                       \
            .opType = ANEURALNETWORKS_##op,                             \
            .supportedDataTypes = {__VA_ARGS__},                        \
            .supportedRanks = {4},                                      \
            .version = HalVersion::ver,                                 \
            .inputs = {INPUT_DEFAULT, PARAMETER_FLOAT_RANGE(0.2, 4.0),  \
                       PARAMETER_FLOAT_RANGE(0.2, 4.0),                 \
                       PARAMETER_CHOICE(Type::BOOL, false, true)},      \
            .outputs = {OUTPUT_DEFAULT},                                \
            .constructor = resizeOpConstructor};

DEFINE_RESIZE_OP_SIGNATURE(RESIZE_BILINEAR, V1_2, Type::TENSOR_FLOAT32, Type::TENSOR_QUANT8_ASYMM,
                           Type::TENSOR_FLOAT16);
DEFINE_RESIZE_OP_SIGNATURE(RESIZE_NEAREST_NEIGHBOR, V1_2, Type::TENSOR_FLOAT32,
                           Type::TENSOR_QUANT8_ASYMM, Type::TENSOR_FLOAT16);

}  // namespace fuzzing_test
}  // namespace nn
}  // namespace android

File: frameworks/rs/tests/java_api/ImageProcessing_jb/src/com/android/rs/image/resize.rs
/*
 * Copyright (C) 2014 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "ip.rsh"
#pragma rs_fp_relaxed

int32_t gWidthIn;
int32_t gHeightIn;
rs_allocation gIn;
float scale;


uchar4 __attribute__((kernel)) nearest(uint32_t x, uint32_t y) {
    float xf = clamp(x * scale, 0.f, (float)gWidthIn - 1.f);
    float yf = clamp(y * scale, 0.f, (float)gHeightIn - 1.f);
    uint32_t ix = xf;
    uint32_t iy = yf;

    uchar4 tmp = rsGetElementAt_uchar4(gIn, ix, iy);
    tmp.a = 0xff;
    return tmp;
}


static float4 cubicInterpolate (float4 p0,float4 p1,float4 p2,float4 p3 , float x) {
    return p1 + 0.5f * x * (p2 - p0 + x * (2.f * p0 - 5.f * p1 + 4.f * p2 - p3
            + x * (3.f * (p1 - p2) + p3 - p0)));
}

uchar4 __attribute__((kernel)) bicubic(uint32_t x, uint32_t y) {
    float xf = (x + 0.5f) * scale - 0.5f;
    float yf = (y + 0.5f) * scale - 0.5f;

    int startx = (int) floor(xf - 1);
    int starty = (int) floor(yf - 1);
    xf = xf - floor(xf);
    yf = yf - floor(yf);
    int maxx = gWidthIn - 1;
    int maxy = gHeightIn - 1;

    uint32_t xs0 = (uint32_t) clamp(startx + 0, 0, maxx);
    uint32_t xs1 = (uint32_t) clamp(startx + 1, 0, maxx);
    uint32_t xs2 = (uint32_t) clamp(startx + 2, 0, maxx);
    uint32_t xs3 = (uint32_t) clamp(startx + 3, 0, maxx);

    uint32_t ys0 = (uint32_t) clamp(starty + 0, 0 , maxy);
    uint32_t ys1 = (uint32_t) clamp(starty + 1, 0 , maxy);
    uint32_t ys2 = (uint32_t) clamp(starty + 2, 0 , maxy);
    uint32_t ys3 = (uint32_t) clamp(starty + 3, 0 , maxy);

    float4 p00 = convert_float4(rsGetElementAt_uchar4(gIn, xs0, ys0));
    float4 p01 = convert_float4(rsGetElementAt_uchar4(gIn, xs1, ys0));
    float4 p02 = convert_float4(rsGetElementAt_uchar4(gIn, xs2, ys0));
    float4 p03 = convert_float4(rsGetElementAt_uchar4(gIn, xs3, ys0));
    float4 p0  = cubicInterpolate(p00, p01, p02, p03, xf);

    float4 p10 = convert_float4(rsGetElementAt_uchar4(gIn, xs0, ys1));
    float4 p11 = convert_float4(rsGetElementAt_uchar4(gIn, xs1, ys1));
    float4 p12 = convert_float4(rsGetElementAt_uchar4(gIn, xs2, ys1));
    float4 p13 = convert_float4(rsGetElementAt_uchar4(gIn, xs3, ys1));
    float4 p1  = cubicInterpolate(p10, p11, p12, p13, xf);

    float4 p20 = convert_float4(rsGetElementAt_uchar4(gIn, xs0, ys2));
    float4 p21 = convert_float4(rsGetElementAt_uchar4(gIn, xs1, ys2));
    float4 p22 = convert_float4(rsGetElementAt_uchar4(gIn, xs2, ys2));
    float4 p23 = convert_float4(rsGetElementAt_uchar4(gIn, xs3, ys2));
    float4 p2  = cubicInterpolate(p20, p21, p22, p23, xf);

    float4 p30 = convert_float4(rsGetElementAt_uchar4(gIn, xs0, ys3));
    float4 p31 = convert_float4(rsGetElementAt_uchar4(gIn, xs1, ys3));
    float4 p32 = convert_float4(rsGetElementAt_uchar4(gIn, xs2, ys3));
    float4 p33 = convert_float4(rsGetElementAt_uchar4(gIn, xs3, ys3));
    float4 p3  = cubicInterpolate(p30, p31, p32, p33, xf);

    float4 p  = cubicInterpolate(p0, p1, p2, p3, yf);
    p = clamp(p + 0.5f, 0.f, 255.f);
    return convert_uchar4(p);
}

